src/Bicep.Core/Diagnostics/DiagnosticTree.cs

```diff
@@ -0,0 +1,146 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using Bicep.Core.Collections.Trees;
+using Bicep.Core.Extensions;
+using Bicep.Core.Parsing;
+using Bicep.Core.PrettyPrint.Documents;
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+using System.Runtime.CompilerServices;
+using System.Text;
+using System.Threading.Tasks;
+using System.Xml;
+
+namespace Bicep.Core.Diagnostics
+{
+    public sealed class DiagnosticTree : IntervalTree<IDiagnostic>, IDiagnosticWriter, IDiagnosticLookup
```

Comment:
I was aiming for adding caching in `ParseDiagnosticsVisitor`, and it eventually evolved into an interval tree. Below is my thought process:

1. To add cache to the `ParseDiagnosticsVisitor`, the visitor itself must be a cached singleton that can be attached to some other class so it can be accessed later.
2. The diagnostic cache in `ParseDiagnosticVisitor` needs to support query diagnostics by syntax, which means the following is true for `HasParseDiagnostics(SyntaxBase)`
    - If a given syntax node `S` has a diagnostic, return true
    - If any of `S`'s descendants has a diagnostic, return true

    And the following is true for `GetParseDiagnostics(SyntaxBase)`:
    - For given syntax node, return `S`'s diagnostic and the diagnostics of its descendants

  
    Thus, a simple HashMap based cache like `IDictionary` or `ILookup` may not be ideal, since we may end up storing many duplicate diagnostics - for a key S, value will have to contain S's diagnostic and its descendants' diagnostics. The fact that the syntax tree is hierarchical suggests that the data structure of the cache may better be a tree. In our case, I think an interval tree can still be used with one modification: instead of searching for intervals overlapping with a target interval, we search for intervals that are enclosed by the target interval. The interval tree `max` can help when querying enclosed intervals, because if a syntax node overlaps with another node, one of them must overlap with the other.

I would like to learn if there are other better data structures that can be used, and I'm open to other better solutions.

Reply:
> Thus, a simple HashMap based cache like IDictionary or ILookup may not be ideal, since we may end up storing many duplicate diagnostics - for a key S, value will have to contain S's diagnostic and its descendants' diagnostics.

Agreed, though I'm not sure how expensive duplicates would be since we'd be duplicating pointers rather than the full content of each diagnostic. I'm not sure if this approach would be efficient, but the visitor could keep a stack of the hierarchy being visited so that each diagnostic encountered could be cached against the syntax node currently being visited plus all of its ancestors.

The one advantage of a hierarchical cache is that it makes it easier to invalidate and recalculate subtrees (if the syntax is modified), but that should be possible with an interval tree as well (so long as there is an easy way to get the syntax node corresponding to an interval -- perhaps if the interval tree stored a reference to the syntax node but used its span for lookups instead of each tree node storing the start and end of the span).

Reply:
Another advantage of the hierarchical cache is that lookup is O(1). I guess I'm not a fan of the space and time complexity it takes to construct the cache, though, since maintaining the stack requires O(n), and we may iterate the stack for every diagnostic we encounter.

Invalidating cache may not be needed as discussed below, but if it is needed, don't we still need to visit a syntax node and its children to get all the keys to delete? Because invaliding a syntax node means all of its descendants should be invalidated as well.

>  that should be possible with an interval tree as well (so long as there is an easy way to get the syntax node corresponding to an interval -- perhaps if the interval tree stored a reference to the syntax node but used its span for lookups instead of each tree node storing the start and end of the span).
Although not used, a `Delete` method for the `DiagnosticTree` is already implemented, which basically queries and delete the corresponding nodes enclosed by a span.

Reply:
> Another advantage of the hierarchical cache is that lookup is O(1). I guess I'm not a fan of the space and time complexity it takes to construct the cache, though, since maintaining the stack requires O(n), and we may iterate the stack for every diagnostic we encounter.

Maintaining the stack shouldn't add any complexity to the visitor, though it would be O(n) to iterate for each diagnostic. I was thinking of something along the lines of:

```csharp
    private readonly Stack<List<IDiagnostic>> currentHierarchyDiagnostics = new();
    private readonly Dictionary<SyntaxBase, ImmutableArray<IDiagnostic>> cache = new();

    protected override void VisitInternal(SyntaxBase node)
    {
        if (cache.TryGetValue(node, out var cached))
        {
            RecordDiagnostics(syntax.Diagnostics);
            return;
        }

        currentHierarchyDiagnostics.Push(new());

        node.Accept(this);

        cache.Add(node, currentHierarchyDiagnostics.Pop().ToImmutableArray());
    }

    public override void VisitSkippedTriviaSyntax(SkippedTriviaSyntax syntax)
    {
        base.VisitSkippedTriviaSyntax(syntax);

        RecordDiagnostics(syntax.Diagnostics);
    }

    private void RecordDiagnostics(IEnumerable<IDiagnostic> diagnostics)
    {
        foreach (var frame in currentHierarchyDiagnostics)
        {
            frame.AddRange(syntax.Diagnostics);
        }
    }
```

> Invalidating cache may not be needed as discussed below, but if it is needed, don't we still need to visit a syntax node and its children to get all the keys to delete? Because invaliding a syntax node means all of its descendants should be invalidated as well.

Yes, that's true. I think you would also need to delete cached values for all of the ancestors of the node if they're meant to store duplicate pointers to the diagnostics of their descendants. Going in both directions seems easier with an interval tree. Though you're right that it seems like invalidation is not needed.

Reply:
Ah, you are right, I missed that we also need to update the cache for ancestors, which adds some complexity.

Btw, as I mentioned in https://github.com/Azure/bicep/pull/10162#discussion_r1141669656, my original goal was to not visit the syntax tree at all to construct the cache and detach diagnostics from `SkippedTriviaSyntax`. but that requires some improvements to the parser, which is out of scope of the PR. I think using the interval tree opens a door to that optimization in the future.

Reply:
Had some time after dinner so went ahead and implemented hierarchical caching. Below are the benchmarks.

### Compilation

#### Hierarchical caching
![Cache](https://user-images.githubusercontent.com/16367959/226788137-9dc6fbfe-daf8-48d2-b547-69570b4f581f.png)

#### Interval tree
![After](https://user-images.githubusercontent.com/16367959/226789447-c788603a-9274-4909-bfb7-9dad607b21db.png)


The latter did slightly better in speed and allocated memory. Hierarchical caching outperforms interval tree in terms of GC.

### Compilation - valid datasets

#### Hierarchical caching
![Valid_Cached](https://user-images.githubusercontent.com/16367959/226788332-6128c95e-baa2-4cd6-a01c-0f7b2d7f88f0.png)

#### Interval tree
![Valid_After](https://user-images.githubusercontent.com/16367959/226789680-277091e8-33a2-4e64-961a-9d9a60c4e5b1.png)

Interval tree did a better job in speed, GC, and allocated memory.

### Compilation - invalid datasets

Didn't take screenshot, but no meaningful differences between the two.

PS: One thing that worth mentioning is that the hierarchical caching is only for parsing errors. The lexing error lookup is still a n interval tree, because a lexing error cannot be attached to a syntax node, and we would have to query them by `TextSpan`. A bonus side effect of that is some of the weird formatter behavior got fixed, for example: https://github.com/Azure/bicep/pull/10162/files?diff=split&w=0#diff-06ebe8df656cb71e1884151c1c44e7b911d1c97853ea35053975dbf30c5907ce